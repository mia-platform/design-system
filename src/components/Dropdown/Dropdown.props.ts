/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { KeyboardEvent, MouseEvent, ReactElement, ReactNode } from 'react'

import { Placement } from './Dropdown.types'
import { Size } from '../../themes/schema/shape'

export enum ItemLayout {
  Horizontal = 'horizontal',
  Vertical = 'vertical'
}

export type DropdownItem = {

  /**
   * renders the icon in a danger fashion, with red tones.
   */
  danger?: boolean,

  /**
   * Optional icon for the menu item
   */
  icon?: ReactNode

  /**
   * The id for the item.
   */
  id: string,

  /**
   * The label to be displayed for the item.
   */
  label: ReactNode,

  /**
   * A secondary label to be displayed for the item.
   */
  secondaryLabel?: ReactNode,

  /**
   * A Tag, displayed on the right side of the item.
   * The argument must be a Tag component, otherwise an error will be thrown.
   */
  tag?: ReactNode

  /**
   * List of sub-menu items.
   */
  children?: DropdownItem[],
}

export enum DropdownTrigger {
  Click = 'click',
  Hover = 'hover',
  ContextMenu = 'contextMenu'
}

/**
 * Each FooterAction is rendered as a button with the provided configuration
 */
export type FooterAction = {

  /**
   * Optional icon
   * */
  icon?: ReactNode

  /**
   * label displayed
   * */
  label?: ReactNode

  /**
   * callback invoked on click.
   */
  onClick: () => void
}

export type DropdownFooter = {

  /**
   * Optional list of actions, they are rendered between the top and bottom nodes.
   */
  actions?: FooterAction[]

  /**
   * Optional node to be rendered at the bottom of the footer.
   */
  bottom?: ReactNode

  /**
   * Optional node to be rendered at the top of the footer.
   */
  top?: ReactNode
}

export type DropdownClickEvent = {

  /**
   * original dom event.
   */
  domEvent: MouseEvent<HTMLElement> | KeyboardEvent<HTMLElement>,

  /**
   * id of the clicked item.
   */
  id: string,

  /**
   * selected item data.
   */
  item?: DropdownItem,

  /**
   * array of strings representing the selected item json path.
   */
  selectedPath: string[],
}

export enum OpenChangeInfoSource {
  Trigger = 'trigger',
  Menu = 'menu'
}

export type OpenChangeInfo = {
  source: OpenChangeInfoSource
}

export type DropdownProps = {

  /**
   * Use this to focus the dropdown element.
   */
  autoFocus?: boolean

  /**
   * The ReactElement used as anchor to trigger the Dropdown visibility.
   */
  children?: ReactElement,

  /**
   * Optional footer configuration.
   */
  footer?: DropdownFooter,

  /**
   * Set to `true` to keep the selected value highlighted after the selection.
   * Useful in case you need to keep the selected option after the dropdown is closed.
   * Defaults to `true`.
   */
  persistSelection?: boolean,

  /**
   * List of items to be shown as selected at first render.
   *
   * _NOTE_: this property does not work if _persistSelection_ is set to `false`
   */
  initialSelectedItems?: string[],

  /**
   * Whether the dropdown menu is disabled.
   */
  isDisabled?: boolean,

  /**
   * list of items to be rendered within the Dropdown.
   */
  items: DropdownItem[],

  /**
   * Allows to control the Dropdown label layout (accepts: horizontal, vertical).
   */
  itemLayout?: ItemLayout,

  /**
   * Sets the max height for the menu item list, this is useful to prevent creating
   * a dropdown with many items that may be too big (default: 240px).
   */
  menuItemsMaxHeight?: Size,

  /**
   * control whether to allow multiple highlight selection.
   *
   * _NOTE_: this property does not work if _persistSelection_ is set to `false`
   */
  multiple?: boolean

  /**
   * @param DropdownClickEvent event contains the reference to the clicked item, specifically it holds
   * - id: the id of the item
   * - selectedPath: array of strings representing the selected item json path
   */
  onClick: (event: DropdownClickEvent) => void

  /**
   * list of triggers that can open the Dropdown (accepts: click, hover, contextMenu).
   */
  triggers?: DropdownTrigger[],

  /**
   * Called when the open state changes. Not triggered when hidden by click item.
   *
   * @param open the current open state of the Dropdown
   * @param info an object containing the `source` of the open change, that can be:
   * - trigger: the open state was changed by an external trigger
   *    (e.g. the dropdown is opening because the children was clicked, or closing because the focus was lost)
   * - menu: the open state was changed by clicking on a menu item
   *    (e.g. the dropdown is closing because a menu item was clicked)
   * @returns
   */
   onOpenChange?: (open: boolean, info: OpenChangeInfo) => void

   /**
    * To set the container of the dropdown menu.
    * The default behavior is to create a div element and append it at the end of the body,
    *  but you can reset it to the scrolling area and make a relative reposition.
    */
   getPopupContainer?: (triggerNode: HTMLElement) => HTMLElement,

    /**
      * The placement of the dropdown menu, one of:
      *
      *  `Placement.TopLeft`, `Placement.TopRight`, `Placement.BottomLeft`,
      *  `Placement.BottomRight`, `Placement.Top`, `Placement.Bottom`
      *
      * Defaults to `Placement.BottomLeft`
      */
   placement?: Placement
}
