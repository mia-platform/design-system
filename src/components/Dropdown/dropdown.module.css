/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

.primaryLabel {
  color: var(--palette-text-neutral-main, #636363);
}

.danger {
  color: var(--palette-action-danger-default, #E50020);
}

.strong {
  font-weight: 600;
}

.secondaryLabel {
  color: var(--palette-text-neutral-subtle, #898989);
}

.dropdownWrapper {
  & :global(.mia-platform-dropdown-menu-item-danger):hover,
  & :global(.mia-platform-dropdown-menu-item-selected.mia-platform-dropdown-menu-item-danger) {
    /* overwrite antd default background */
    background-color: var(--palette-action-danger-hover, #FF5453) !important;
  }

  & :global(.mia-platform-dropdown-menu-item-danger):hover .danger,
  & :global(.mia-platform-dropdown-menu-item-selected) .danger{
    color: var(--palette-common-white, #ffffff);
  }

  & :global(.mia-platform-dropdown-menu-item-selected) .primaryLabel {
    font-weight: var(--typography-bodyLBold-fontWeight, 600);
  }

  :global(.mia-platform-dropdown-menu-submenu-title) {
    display: flex !important;
  }

  :global(.mia-platform-dropdown-menu-submenu) {
    background: none !important;
    box-shadow: none !important;
  }

  :global(div.mia-platform-dropdown-menu-submenu) {
    max-width: fit-content;
  }

  :global(ul.mia-platform-dropdown-menu) {
    /* overwrite default radius defined by our theme */
    border-bottom-left-radius: var(--shape-border-radius-lg, 8px);
    border-bottom-right-radius: var(--shape-border-radius-lg, 8px);
    border-top-left-radius: var(--shape-border-radius-lg, 8px);
    border-top-right-radius: var(--shape-border-radius-lg, 8px);
  }

  :global(.mia-platform-tag) {
    margin-inline-end: unset;
  }
}

.dropdownWithFooterRenderWrapper {
  background-color: white;
  width: 100%;
  border-radius: var(--shape-border-radius-lg, 8px);
  gap: var(--spacing-gap-xs, 4px)
}

.itemRowContainer {
  display: flex;
  flex-direction: row;
  gap: var(--spacing-gap-sm, 8px);
  align-items: center;
}

.labelsContainer {
  display: flex;
  text-align: left;
  flex: 1 1 auto;
}

.horizontalContainer {
  flex-direction: row;
  gap: var(--spacing-gap-sm, 8px);
  align-items: center;
}

.verticalContainer {
  flex-direction: column;
  gap: var(--spacing-gap-xs, 4px);
}
