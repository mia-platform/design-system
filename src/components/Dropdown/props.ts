/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { KeyboardEvent, MouseEvent, ReactElement, ReactNode } from 'react'

import { Placement } from './types'
import { Size } from '../../themes/schema/shape'

export enum ItemLayout {
  Horizontal = 'horizontal',
  Vertical = 'vertical'
}

export type DropdownItem = {

  /**
   * renders the icon in a danger fashion, with red tones.
   */
  danger?: boolean,

  /**
   * Optional icon for the menu item
   */
  icon?: ReactNode

  /**
   * The id for the item.
   */
  id: string,

  /**
   * The label to be displayed for the item.
   *
   * When the automatic dropdown search is enabled via the `isSearchable` prop,
   * it will only filter items with a **string** or **number** label.
   * For labels that are ReactNodes, filtering must be handled externally via the `onSearch` callback.
   */
  label: ReactNode,

  /**
   * A secondary label to be displayed for the item.
   */
  secondaryLabel?: ReactNode,

  /**
   * A Tag, displayed on the right side of the item.
   * The argument must be a Tag component, otherwise an error will be thrown.
   */
  tag?: ReactNode

  /**
   * Whether menu item is disabled
   */
  disabled?: boolean

  /**
   * List of sub-menu items.
   */
  children?: DropdownItem[],
}

export enum DropdownTrigger {
  Click = 'click',
  Hover = 'hover',
  ContextMenu = 'contextMenu'
}

/**
 * Each FooterAction is rendered as a button with the provided configuration
 */
export type FooterAction = {

  /**
   * Optional icon
   * */
  icon?: ReactNode

  /**
   * label displayed
   * */
  label?: ReactNode

  /**
   * callback invoked on click.
   */
  onClick: () => void
}

export type DropdownHeader = {

  /**
   * Content displayed **above** the search input (if visible via `isSerchable` prop).
   * if the search input is not visible should be provided only one of `top` or `bottom` props.
   */
  top?: ReactNode

  /**
   * Content displayed **below** the search input (if visible via `isSerchable` prop).
   * if the search input is not visible should be provided only one of `top` or `bottom` props.
   */
  bottom?: ReactNode
}

export type DropdownFooter = {

  /**
   * Optional list of actions, they are rendered between the top and bottom nodes.
   */
  actions?: FooterAction[]

  /**
   * Optional node to be rendered at the bottom of the footer.
   */
  bottom?: ReactNode

  /**
   * Optional node to be rendered at the top of the footer.
   */
  top?: ReactNode
}

export type DropdownClickEvent = {

  /**
   * original dom event.
   */
  domEvent: MouseEvent<HTMLElement> | KeyboardEvent<HTMLElement>,

  /**
   * id of the clicked item.
   */
  id: string,

  /**
   * selected item data.
   */
  item?: DropdownItem,

  /**
   * array of strings representing the selected item json path.
   */
  selectedPath: string[],
}

export enum OpenChangeInfoSource {
  Trigger = 'trigger',
  Menu = 'menu'
}

export type OpenChangeInfo = {
  source: OpenChangeInfoSource
}

export type DropdownProps = {

  /**
   * Use this to focus the dropdown element.
   */
  autoFocus?: boolean

  /**
   * The ReactElement used as anchor to trigger the Dropdown visibility.
   */
  children?: ReactElement

  /**
   * Optional header configuration.
   */
  header?: DropdownHeader

  /**
   * Optional footer configuration.
   */
  footer?: DropdownFooter

  /**
   * Set to `true` to keep the selected value highlighted after the selection.
   * Useful in case you need to keep the selected option after the dropdown is closed.
   * Defaults to `true`.
   */
  persistSelection?: boolean

  /**
   * List of items to be shown as selected at first render.
   *
   * _NOTE_: this property does not work if _persistSelection_ is set to `false`
   */
  initialSelectedItems?: string[]

  /**
   * Whether the dropdown menu is disabled.
   */
  isDisabled?: boolean

  /**
   * list of items to be rendered within the Dropdown.
   */
  items: DropdownItem[]

  /**
   * Allows to control the Dropdown label layout (accepts: horizontal, vertical).
   */
  itemLayout?: ItemLayout

  /**
   * Sets the max height for the menu item list, this is useful to prevent creating
   * a dropdown with many items that may be too big (default: 240px).
   */
  menuItemsMaxHeight?: Size

  /**
   * control whether to allow multiple highlight selection.
   *
   * _NOTE_: this property does not work if _persistSelection_ is set to `false`
   */
  multiple?: boolean

  /**
   * @param DropdownClickEvent event contains the reference to the clicked item, specifically it holds
   * - id: the id of the item
   * - selectedPath: array of strings representing the selected item json path
   */
  onClick: (event: DropdownClickEvent) => void

  /**
   * list of triggers that can open the Dropdown (accepts: click, hover, contextMenu).
   */
  triggers?: DropdownTrigger[]

  /**
   * Called when the open state changes. Not triggered when hidden by click item.
   *
   * @param open the current open state of the Dropdown
   * @param info an object containing the `source` of the open change, that can be:
   * - trigger: the open state was changed by an external trigger
   *    (e.g. the dropdown is opening because the children was clicked, or closing because the focus was lost)
   * - menu: the open state was changed by clicking on a menu item
   *    (e.g. the dropdown is closing because a menu item was clicked)
   * @returns
   */
  onOpenChange?: (open: boolean, info: OpenChangeInfo) => void

  /**
   * To set the container of the dropdown menu.
   * The default behavior is to create a div element and append it at the end of the body,
   *  but you can reset it to the scrolling area and make a relative reposition.
   */
  getPopupContainer?: (triggerNode: HTMLElement) => HTMLElement

  /**
   * The placement of the dropdown menu, one of:
   *
   *  `Placement.TopLeft`, `Placement.TopRight`, `Placement.BottomLeft`,
   *  `Placement.BottomRight`, `Placement.Top`, `Placement.Bottom`
   *
   * Defaults to `Placement.BottomLeft`
   */
  placement?: Placement

  /**
   * Enables a search input for filtering items in the component.
   *
   * - If `onSearch` is **not** provided, the component applies its own
   *   built-in filtering (case-insensitive substring match on string labels).
   * - If `onSearch` **is** provided, filtering is delegated to the parent
   *   component via the callback.
   */
  isSearchable?: boolean

  /**
   * Called when the user submits a search query.
   *
   * Providing this prop overrides the component's built-in filtering logic,
   * allowing the parent component to handle search behavior.
   *
   * @param query - The text entered by the user into the search input.
   */
  onSearch?: (query: string) => void

  /**
   * Placeholder text to show inside the search input.
   * Default to 'Search...'
   */
  searchPlaceholder?: string

  /**
   * Whether the component should show a loading state.
   */
  isLoading?: boolean

  /**
   * Whether the component should show an error state.
   */
  hasError?: boolean

  /**
   * Error message to display when `hasError` is true.
   * Default to 'An error occurred'
   */
  errorMessage?: string

  /**
   * Callback fired when the user retries after an error.
   *
   * @param query - The last search query string, if available.
   */
  onRetry?: (query: string) => void

  /**
   * Enables infinite scrolling functionality for the dropdown.
   * When enabled, the `onScrollEndReached` callback will be triggered
   * when the user scrolls near the bottom of the items list.
   */
  isInfiniteScrollEnabled?: boolean

  /**
   * Callback fired when the user scrolls near the bottom of the dropdown items.
   * This is typically used to load more items for infinite scrolling.
   * Only triggered when `isInfiniteScrollEnabled` is true.
   */
  onScrollEndReached?: () => void

  /**
   * Additional items to append to the existing dropdown items.
   * These items are typically loaded asynchronously and added to the dropdown
   * when infinite scrolling loads more data.
   */
  incrementalItems?: DropdownItem[]

  /**
   * Indicates whether additional items are currently being loaded.
   * When true, a loading skeleton is displayed at the bottom of the dropdown
   * to provide visual feedback during the loading process.
   */
  isLoadingIncrementalItems?: boolean

  /**
   * Distance in pixels from the bottom of the scrollable area
   * that triggers the `onScrollEndReached` callback.
   * Defaults to 32 pixels.
   */
  scrollThreshold?: number
};
