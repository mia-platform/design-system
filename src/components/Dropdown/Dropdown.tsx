/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { Dropdown as AntdDropdown, type MenuProps as AntdMenuProps } from 'antd'
import React, { ReactElement, ReactNode, useCallback, useMemo } from 'react'

import { DropdownClickEvent, DropdownItem, DropdownProps, DropdownTrigger, ItemLayout } from './props'
import Label from './components/Label/Label'
import styles from './dropdown.module.css'

type ArrayElement<ArrayType extends readonly unknown[] | undefined> =
  ArrayType extends readonly (infer ElementType)[] ? ElementType : never;

type AntdMenuItems = AntdMenuProps['items']
type AntdMenuItem = ArrayElement<AntdMenuItems>

type AntdMenuClickEvent = {
  key: string,
  keyPath: string[],
  domEvent: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLElement>
}

export const defaults = {
  itemLayout: ItemLayout.Horizontal,
  trigger: [DropdownTrigger.Click],
}

const itemMatcher = (id: string) => (item: DropdownItem): boolean => item.id === id

export const Dropdown = ({
  autoFocus,
  children,
  isDisabled,
  itemLayout = defaults.itemLayout,
  items,
  onClick,
  triggers = defaults.trigger,
}: DropdownProps): ReactElement => {
  const findItem = useCallback((id: string) => items.find(itemMatcher(id)), [items])

  const antdItems = useMemo<AntdMenuItems>(() => itemsAdapter(items, itemLayout), [itemLayout, items])
  const innerNode = useMemo(() => (children ? <span>{children}</span> : null), [children])

  const onAntdMenuClick = useCallback(
    (antdEvent: AntdMenuClickEvent) => onClick(eventAdapter(antdEvent, findItem)),
    [findItem, onClick]
  )

  const dropdownRender = useCallback((menu: ReactNode): ReactNode => {
    return React.cloneElement(menu as ReactElement)
  }, [])

  const menu = useMemo(() => ({
    items: antdItems,
    onClick: onAntdMenuClick,
  }), [antdItems, onAntdMenuClick])

  return (
    <AntdDropdown
      autoFocus={autoFocus}
      disabled={isDisabled}
      dropdownRender={dropdownRender}
      menu={menu}
      overlayClassName={styles.dropdownWrapper}
      trigger={triggers}
    >
      {innerNode}
    </AntdDropdown>
  )
}

Dropdown.ItemLayout = ItemLayout
Dropdown.Trigger = DropdownTrigger

function itemsAdapter(items: DropdownItem[], layout: ItemLayout): AntdMenuItems {
  return items.map<AntdMenuItem>((item: DropdownItem) => ({
    label: <Label item={item} layout={layout} />,
    key: item.id,
    danger: item.danger,
  }))
}

function eventAdapter(
  event: AntdMenuClickEvent,
  itemFinder: (id: string) => DropdownItem|undefined,
): DropdownClickEvent {
  return {
    id: event.key,
    selectedPath: event.keyPath,
    domEvent: event.domEvent,
    item: itemFinder(event.key),
  }
}
