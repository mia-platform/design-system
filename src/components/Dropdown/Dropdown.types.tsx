import { type MenuProps as AntdMenuProps } from 'antd'

import { OpenChangeInfoSource } from './Dropdown.props'

export type ArrayElement<ArrayType extends readonly unknown[] | undefined> =
ArrayType extends readonly (infer ElementType)[] ? ElementType : never;

export type AntdMenuItems = AntdMenuProps['items']
export type AntdMenuItem = ArrayElement<AntdMenuItems>

export type AntdMenuClickEvent = {
   key: string,
   keyPath: string[],
   domEvent: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLElement>
}

export type AntdTriggerSource = 'trigger' | 'menu'
export const antdSourceMap: Record<AntdTriggerSource, OpenChangeInfoSource> = {
  'trigger': OpenChangeInfoSource.Trigger,
  'menu': OpenChangeInfoSource.Menu,
}

export enum Placement {
  TopLeft = 'topLeft',
  TopRight = 'topRight',
  BottomLeft = 'bottomLeft',
  BottomRight = 'bottomRight',
  Top = 'top',
  Bottom = 'bottom',
}

