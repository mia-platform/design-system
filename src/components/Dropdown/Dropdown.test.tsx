/* eslint-disable max-nested-callbacks */
/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { DropdownItem, DropdownProps, DropdownTrigger, OpenChangeInfoSource } from './props'
import { RenderResult, render, screen, userEvent, waitFor } from '../../test-utils'
import { Button } from '../Button'
import { Dropdown } from './Dropdown'

const items: DropdownItem[] = [
  { id: '1', label: 'Label 1' },
  { id: '2', label: 'Label 2', secondaryLabel: 'Additional Info 2' },
  { id: '3', label: 'Danger Label', secondaryLabel: 'Additional Info 2', danger: true },
]
const defaultProps: DropdownProps = {
  items,
  children: <Button>{'test-trigger-button'}</Button>,
  onClick: jest.fn(),
}

describe('Dropdown Component', () => {
  beforeEach(() => jest.clearAllMocks())

  describe('triggers', () => {
    it('opens dropdown on click', async() => {
      renderDropdown()
      const button = screen.getByText('test-trigger-button')
      userEvent.click(button)
      await screen.findByRole('menuitem', { name: 'Label 1' })
    })

    it('opens dropdown on hover', async() => {
      renderDropdown({ props: { ...defaultProps, triggers: [DropdownTrigger.Hover] } })
      const button = screen.getByText('test-trigger-button')
      userEvent.hover(button)
      await screen.findByRole('menuitem', { name: 'Label 1' })
    })
  })

  describe('label layouts', () => {
    it('render labels', async() => {
      renderDropdown({ props: { ...defaultProps, items } })
      const button = screen.getByText('test-trigger-button')
      userEvent.click(button)

      await screen.findByRole('menuitem', { name: 'Label 1' })

      expect(screen.getAllByRole('menuitem')).toHaveLength(3)

      const [first, second, third] = screen.getAllByRole('menuitem')
      expect(first).toMatchSnapshot()
      expect(second).toMatchSnapshot()
      expect(third).toMatchSnapshot()
    })
  })

  describe('onClick', () => {
    it('invokes onClick with correct id', async() => {
      const onClick = jest.fn()
      const props = {
        ...defaultProps,
        onClick,
      }
      renderDropdown({ props })
      const button = screen.getByText('test-trigger-button')
      userEvent.click(button)

      const item = await screen.findByRole('menuitem', { name: 'Label 1' })
      userEvent.click(item)

      await waitFor(() => expect(onClick).toHaveBeenCalled())
      expect(onClick).toHaveBeenCalledTimes(1)
      const [[invocation]] = onClick.mock.calls
      expect(invocation.id).toEqual('1')
      expect(invocation.selectedPath).toEqual(['1'])
      expect(invocation.item).toEqual(items[0])
    })

    describe('nesting', () => {
      it('invokes onClick with correct first-level nested id', async() => {
        const onClick = jest.fn()
        const caseItems = [
          { id: '1', label: 'Label 1' },
          {
            id: '2',
            label: 'Label 2',
            children: [{
              id: '2-1',
              label: 'Label 2-1',
            }, {
              id: '2-2',
              label: 'Label 2-2',
              children: [{
                id: '2-2-1',
                label: 'Label 2-2-1',
              }],
            }],
          },
        ]
        const props = {
          ...defaultProps,
          items: caseItems,
          onClick,
        }
        renderDropdown({ props })
        const button = screen.getByText('test-trigger-button')
        userEvent.click(button)

        const item = await screen.findByRole('menuitem', { name: 'Label 1' })
        userEvent.click(item)

        // eslint-disable-next-line max-nested-callbacks
        await waitFor(() => expect(onClick).toHaveBeenCalled())
        expect(onClick).toHaveBeenCalledTimes(1)
        const [[invocation]] = onClick.mock.calls
        expect(invocation.id).toEqual('1')
        expect(invocation.selectedPath).toEqual(['1'])
        expect(invocation.item).toEqual(items[0])

        onClick.mockClear()

        userEvent.click(button)
        const item2 = await screen.findByRole('menuitem', { name: /^Label 2/i })
        userEvent.hover(item2)

        const sub1 = await screen.findByRole('menuitem', { name: 'Label 2-1' }, { timeout: 10000 })
        expect(sub1).toBeInTheDocument()

        userEvent.click(sub1)
        // eslint-disable-next-line max-nested-callbacks
        await waitFor(() => expect(onClick).toHaveBeenCalled())
        expect(onClick).toHaveBeenCalledTimes(1)
        const [[invocation2]] = onClick.mock.calls
        expect(invocation2.id).toEqual('2-1')
        expect(invocation2.selectedPath).toEqual(['2', '2-1'])
        expect(invocation2.item).toEqual(caseItems[1].children![0])
      })

      it('invokes onClick with correct second-level nested id', async() => {
        const onClick = jest.fn()
        const caseItems = [
          { id: '1', label: 'Label 1' },
          {
            id: '2',
            label: 'Label 2',
            children: [{
              id: '2-1',
              label: 'Label 2-1',
            }, {
              id: '2-2',
              label: 'Label 2-2',
              children: [{
                id: '2-2-1',
                label: 'Label 2-2-1',
              }],
            }],
          },
        ]
        const props = {
          ...defaultProps,
          items: caseItems,
          onClick,
        }
        renderDropdown({ props })
        const button = screen.getByText('test-trigger-button')
        userEvent.click(button)

        await screen.findByRole('menuitem', { name: 'Label 1' })
        userEvent.hover(screen.getByRole('menuitem', { name: /^Label 2/i }))
        await screen.findByRole('menuitem', { name: 'Label 2-1' }, { timeout: 10000 })
        userEvent.hover(screen.getByRole('menuitem', { name: 'Label 2-2 right' }))

        const sub1 = await screen.findByRole('menuitem', { name: 'Label 2-2-1' }, { timeout: 10000 })
        expect(sub1).toBeInTheDocument()

        userEvent.click(sub1)
        // eslint-disable-next-line max-nested-callbacks
        await waitFor(() => expect(onClick).toHaveBeenCalled())
        expect(onClick).toHaveBeenCalledTimes(1)
        const [[invocation]] = onClick.mock.calls
        expect(invocation.id).toEqual('2-2-1')
        expect(invocation.selectedPath).toEqual(['2', '2-2', '2-2-1'])
        expect(invocation.item).toEqual(caseItems[1].children![1].children![0])
      }, 20000)
    })
  })

  describe('onOpenChange', () => {
    it('invokes onOpenChange with trigger source when opening', async() => {
      const onOpenChange = jest.fn()
      const props = {
        ...defaultProps,
        onOpenChange,
      }

      renderDropdown({ props })
      const button = screen.getByText('test-trigger-button')
      userEvent.click(button)

      await waitFor(() => expect(onOpenChange).toHaveBeenCalledTimes(1))
      expect(onOpenChange).toHaveBeenCalledWith(true, { source: OpenChangeInfoSource.Trigger })
    })

    it('invokes onOpenChange with menu source when closing clicking on a menu item', async() => {
      const onOpenChange = jest.fn()
      const props = {
        ...defaultProps,
        onOpenChange,
      }

      renderDropdown({ props })
      const button = screen.getByText('test-trigger-button')
      userEvent.click(button)

      await screen.findByRole('menuitem', { name: 'Label 1' })
      userEvent.click(screen.getByRole('menuitem', { name: 'Label 1' }))

      await waitFor(() => expect(onOpenChange).toHaveBeenCalledTimes(2))
      expect(onOpenChange).toHaveBeenNthCalledWith(2, false, { source: OpenChangeInfoSource.Menu })
    })
  })

  describe('highlights selectedItems', () => {
    describe('single mode', () => {
      it('renders proper highlight', async() => {
        const props: DropdownProps = {
          ...defaultProps,
          initialSelectedItems: ['1'],
        }

        renderDropdown({ props })
        const button = screen.getByText('test-trigger-button')
        userEvent.click(button)
        const el = await screen.findByRole('menuitem', { name: 'Label 1' })
        expect(el).toMatchSnapshot()
      })

      it('updates highlight', async() => {
        const onClick = jest.fn()
        const props: DropdownProps = {
          ...defaultProps,
          onClick,
          initialSelectedItems: ['1'],
        }

        renderDropdown({ props })
        const button = screen.getByText('test-trigger-button')
        userEvent.click(button)
        const first = await screen.findByRole('menuitem', { name: 'Label 1' })
        expect(first).toMatchSnapshot('at first render Label 1 is selected')

        const second = screen.getByRole('menuitem', { name: /Label 2/ })
        userEvent.click(second)

        await waitFor(() => expect(onClick).toHaveBeenCalled())

        userEvent.click(button)
        const firstUpdated = await screen.findByRole('menuitem', { name: 'Label 1' })
        const secondUpdated = screen.getByRole('menuitem', { name: /Label 2/ })
        expect(firstUpdated).toMatchSnapshot('after click render Label 1 is not selected')
        expect(secondUpdated).toMatchSnapshot('after click render Label 2 is selected')
      })
    })
  })
})

function renderDropdown(
  { props }: { props: DropdownProps } = { props: defaultProps }
): RenderResult {
  return render(<Dropdown {...props} />)
}
