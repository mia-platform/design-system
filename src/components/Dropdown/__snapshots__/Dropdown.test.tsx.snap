// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Component highlights selectedItems renders proper highlight 1`] = `
<li
  class="mia-platform-dropdown-menu-item mia-platform-dropdown-menu-item-selected mia-platform-dropdown-menu-item-only-child"
  data-menu-id="rc-menu-uuid-test-1"
  role="menuitem"
  tabindex="-1"
>
  <span
    class="mia-platform-dropdown-menu-title-content"
  >
    <div
      class="labelContainer horizontalContainer"
    >
      <span
        class="primaryLabel"
      >
        Label 1
      </span>
    </div>
  </span>
</li>
`;

exports[`Dropdown Component label layouts render labels 1`] = `
<li
  class="mia-platform-dropdown-menu-item mia-platform-dropdown-menu-item-only-child"
  data-menu-id="rc-menu-uuid-test-1"
  role="menuitem"
  tabindex="-1"
>
  <span
    class="mia-platform-dropdown-menu-title-content"
  >
    <div
      class="labelContainer horizontalContainer"
    >
      <span
        class="primaryLabel"
      >
        Label 1
      </span>
    </div>
  </span>
</li>
`;

exports[`Dropdown Component label layouts render labels 2`] = `
<li
  class="mia-platform-dropdown-menu-item mia-platform-dropdown-menu-item-only-child"
  data-menu-id="rc-menu-uuid-test-2"
  role="menuitem"
  tabindex="-1"
>
  <span
    class="mia-platform-dropdown-menu-title-content"
  >
    <div
      class="labelContainer horizontalContainer"
    >
      <span
        class="primaryLabel"
      >
        Label 2
      </span>
      <span
        class="secondaryLabel"
      >
        ·
      </span>
      <span
        class="secondaryLabel"
      >
        Additional Info 2
      </span>
    </div>
  </span>
</li>
`;

exports[`Dropdown Component label layouts render labels 3`] = `
<li
  class="mia-platform-dropdown-menu-item mia-platform-dropdown-menu-item-danger mia-platform-dropdown-menu-item-only-child"
  data-menu-id="rc-menu-uuid-test-3"
  role="menuitem"
  tabindex="-1"
>
  <span
    class="mia-platform-dropdown-menu-title-content"
  >
    <div
      class="labelContainer horizontalContainer"
    >
      <span
        class="primaryLabel danger"
      >
        Danger Label
      </span>
      <span
        class="secondaryLabel danger"
      >
        ·
      </span>
      <span
        class="secondaryLabel danger"
      >
        Additional Info 2
      </span>
    </div>
  </span>
</li>
`;
